File: src/data/courses/dsa/subarrays-subsets-recursion/backtracking/palindrome-partitioning/code/python/palindrome-partitioning.py
# Time: O(N × 2^N), Space: O(N)

def partition(s):
    result = []
    
    def is_palindrome(left, right):
        while left < right:
            if s[left] != s[right]:
                return False
            left += 1
            right -= 1
        return True
    
    def backtrack(index, current):
        if index == len(s):
            result.append(current[:])
            return
        
        for end in range(index, len(s)):
            if is_palindrome(index, end):
                current.append(s[index:end+1])
                backtrack(end + 1, current)
                current.pop()
    
    backtrack(0, [])
    return result

s = "aab"
print(partition(s))

File: src/data/courses/dsa/subarrays-subsets-recursion/backtracking/palindrome-partitioning/code/java/PalindromePartitioning.java
// Time: O(N × 2^N), Space: O(N)

import java.util.*;

class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> result = new ArrayList<>();
        backtrack(s, 0, new ArrayList<>(), result);
        return result;
    }
    
    private void backtrack(String s, int index, List<String> current, List<List<String>> result) {
        if (index == s.length()) {
            result.add(new ArrayList<>(current));
            return;
        }
        
        for (int end = index; end < s.length(); end++) {
            if (isPalindrome(s, index, end)) {
                current.add(s.substring(index, end + 1));
                backtrack(s, end + 1, current, result);
                current.remove(current.size() - 1);
            }
        }
    }
    
    private boolean isPalindrome(String s, int left, int right) {
        while (left < right) {
            if (s.charAt(left++) != s.charAt(right--)) return false;
        }
        return true;
    }
}

public class PalindromePartitioning {
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.partition("aab"));
    }
}
